File: app/src/main/java/com/cbr/behancesampleapp/ui/common/mvp/BaseMvpPresenter.java
package com.cbr.behancesampleapp.ui.common.mvp;

import io.reactivex.disposables.CompositeDisposable;

/** Created by Dimitrios on 8/26/2017.*/
public abstract class BaseMvpPresenter<VIEW extends MvpView> implements MvpPresenter {

	private CompositeDisposable mCompositeDisposable;

	private VIEW mMvpView;

	public BaseMvpPresenter(VIEW view) {
		this.mMvpView = view;
	}

	public VIEW getMvpView() {
		return mMvpView;
	}

	public CompositeDisposable getCompositeDisposable() {
		return mCompositeDisposable;
	}

	@Override
	public void subscribe() {
		mCompositeDisposable = new CompositeDisposable();
	}

	@Override
	public void unsubscribe() {
		cancelPending();
		mCompositeDisposable.dispose();
	}

	public void cancelPending() {
		if (mCompositeDisposable != null && !mCompositeDisposable.isDisposed()) {
			mCompositeDisposable.clear();
		}
	}
}

File: app/src/main/java/com/cbr/behancesampleapp/ui/common/mvp/BaseMvpPresenter.kt
package com.cbr.behancesampleapp.ui.common.mvp

import com.cbr.behancesampleapp.ui.common.MvpPresenter
import com.cbr.behancesampleapp.ui.common.MvpView
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

open class BasePresenter<VIEW : MvpView> constructor(val compositeDisposable: CompositeDisposable) : MvpPresenter<VIEW> {

    lateinit var weakReference: WeakReference<VIEW>

    val view: VIEW?
        get() = weakReference.get()

    override fun attachView(view: VIEW) {
        weakReference = WeakReference(view)
    }

    override fun onSubscribe() {

    }


    override fun onUnsubscribe() {
        cancelPending()
    }

    fun cancelPending() {
        if (!compositeDisposable.isDisposed) {
            compositeDisposable.clear()
        }
    }

}